drivers/char/rust_example.rs

```diff
@@ -51,23 +53,125 @@ module! {
     },
 }
 
-struct RustFile;
+const MAX_TOKENS: usize = 3;
 
-impl FileOperations for RustFile {
+struct SharedStateInner {
+    token_count: usize,
+}
+
+struct SharedState {
+    state_changed: CondVar,
+    inner: Mutex<SharedStateInner>,
+}
+
+impl SharedState {
+    fn try_new() -> KernelResult<Arc<Self>> {
```

Comment:
@alex -- this is an example of the ugliness brought by `Pin` I mentioned in one of the meetings. A bunch of `unsafe` that wouldn't need to be unsafe if we had better support for pinning. Two aspects are relevant here:
1. Initialisation: `CondVar::new` and `Mutex::new` are unsafe because they require a call to `init` after the condvar/mutex are in their final pinned location.
2. Projection: the compiler can't (or won't) verify for us that we don't move out of the fields of a pinned structure, so we need unsafe. (Though admittedly I don't explicitly say that `state` is pinned, mostly because it doesn't give me anything.)

It would be nice if we could avoid this.

Reply:
Right. When you say "we had better support for pinning", do you think this is something that Rust-for-Linux can improve itself, or do you think this requires better language level support?

Reply:
Well, we could choose to ignore `Pin` and assume that everything that is heap-allocated is pinned, but my opinion is that this isn't desirable because `Pin` is actually useful. (We sort of discussed this in the meeting.)

So I think we need better language support.

Reply:
@joshtriplett as our resident friend upstream, what's the best way for us to share these observations/desires with the team?

Reply:
@RalfJung -- since you commented on something else related to this project, I thought I'd ask you if you have any thoughts about improving pinning support. As it is today, it forces us (unreasonably in my opinion) to use too much unsafe code.

Reply:
> Projection: the compiler can't (or won't) verify for us that we don't move out of the fields of a pinned structure, so we need unsafe. (Though admittedly I don't explicitly say that state is pinned, mostly because it doesn't give me anything.)

The pin project crate has a proc macro to help with this.

Reply:
I agree that writing code with pin projections currently requires more unsafe code than it should. I am not sure what the best way forward for that is; the pin-project crate mentioned by @bjorn3 is a great start showing how far one can go even without any language support.

The "new and then init" pattern seems mostly orthogonal to pinning; that pattern inherently requires exposing the user to an unsafe pre-init state. If you want to expose that state safely, you need to give it a separate type.

Reply:
> since you commented on something else related to this project, I thought I'd ask you

Sure, feel free to. :) I don't have a ton of time these days, but I am *super excited* about Rust having a realistic chance of being used in the upstream Linux kernel, so if I can help in small ways I'm more than happy to do that.